envbuilder.go

```diff
@@ -586,117 +652,130 @@ func Run(ctx context.Context, opts options.Options) error {
+	// overridden by the image and devcontainer settings.
+	err = setEnvFromEtcEnvironment()
+	if err != nil {
+		return fmt.Errorf("set env from /etc/environment: %w", err)
+	}
 
-		// Set Envbuilder runtime markers
-		containerEnv["ENVBUILDER"] = "true"
-		if devcontainerPath != "" {
-			containerEnv["DEVCONTAINER"] = "true"
-			containerEnv["DEVCONTAINER_CONFIG"] = devcontainerPath
-		}
+	allEnvKeys := make(map[string]struct{})
 
-		for _, env := range []map[string]string{containerEnv, remoteEnv} {
-			envKeys := make([]string, 0, len(env))
-			for key := range env {
-				envKeys = append(envKeys, key)
-				allEnvKeys[key] = struct{}{}
-			}
-			sort.Strings(envKeys)
-			for _, envVar := range envKeys {
-				value := devcontainer.SubstituteVars(env[envVar], opts.WorkspaceFolder, os.LookupEnv)
-				os.Setenv(envVar, value)
-			}
-		}
+	// It must be set in this parent process otherwise nothing will be found!
+	for _, env := range runtimeData.ImageEnv {
+		pair := strings.SplitN(env, "=", 2)
+		os.Setenv(pair[0], pair[1])
+		allEnvKeys[pair[0]] = struct{}{}
+	}
 
-		// Do not export env if we skipped a rebuild, because ENV directives
-		// from the Dockerfile would not have been processed and we'd miss these
-		// in the export. We should have generated a complete set of environment
-		// on the intial build, so exporting environment variables a second time
-		// isn't useful anyway.
-		if opts.ExportEnvFile != "" && !skippedRebuild {
-			exportEnvFile, err := os.Create(opts.ExportEnvFile)
-			if err != nil {
-				return fmt.Errorf("failed to open EXPORT_ENV_FILE %q: %w", opts.ExportEnvFile, err)
-			}
+	// Set Envbuilder runtime markers
+	runtimeData.ContainerEnv["ENVBUILDER"] = "true"
+	if runtimeData.DevcontainerPath != "" {
+		runtimeData.ContainerEnv["DEVCONTAINER"] = "true"
+		runtimeData.ContainerEnv["DEVCONTAINER_CONFIG"] = runtimeData.DevcontainerPath
+	}
 
-			envKeys := make([]string, 0, len(allEnvKeys))
-			for key := range allEnvKeys {
-				envKeys = append(envKeys, key)
-			}
-			sort.Strings(envKeys)
-			for _, key := range envKeys {
-				fmt.Fprintf(exportEnvFile, "%s=%s\n", key, os.Getenv(key))
-			}
+	for _, env := range []map[string]string{runtimeData.ContainerEnv, runtimeData.RemoteEnv} {
+		envKeys := make([]string, 0, len(env))
+		for key := range env {
+			envKeys = append(envKeys, key)
+			allEnvKeys[key] = struct{}{}
+		}
+		sort.Strings(envKeys)
+		for _, envVar := range envKeys {
+			value := devcontainer.SubstituteVars(env[envVar], opts.WorkspaceFolder, os.LookupEnv)
+			os.Setenv(envVar, value)
+		}
+	}
 
-			exportEnvFile.Close()
+	// Do not export env if we skipped a rebuild, because ENV directives
+	// from the Dockerfile would not have been processed and we'd miss these
+	// in the export. We should have generated a complete set of environment
+	// on the intial build, so exporting environment variables a second time
+	// isn't useful anyway.
+	if opts.ExportEnvFile != "" && !runtimeData.SkippedRebuild {
```

Comment:
nit: I believe this is a fundamental problem with this script. It would be cool to have `runtimeData` is immutable, and use helper functions instead `runtimeData.IsBuildSkipped()`

Reply:
I'm not sure how to achieve immutability here. For this particular function I suppose you could stat the file every time you want to use the bool, but seems like needless work to hit the fs each time. It doesn't work for most of the other data though.